### DISCUSSION TITLE
Custom Nixie HW requirements

### DISCUSSION THREAD

**Discussion by orrious** [tagyoureit/nodejs-poolController#507] [Nixie]
In https://github.com/tagyoureit/nodejs-poolController/discussions/497 we found that there really isn't an ideal solution for my hardware needs.  I try not to do anything just for myself, so towards that end I'd like to capture what the community would like to see included in one ore more boards.

Lets create an "answer" per input/output.

**Comment by orrious** [comment]
Bus protocol:  I2C / SPI / RS485 / Ethernet / Wifi

**Reply by orrious** [reply]
I2C - If we create a "hat" style board for a Raspberry PI
RS485 (MODBUS) - If we create standalone boards.  I would recommend a dedicated RS485 bus for this use.  It would also allow for very reliable long distance remote control via off the shelf ESP32 / LoraWAN radios.

Given the high demand / low availability of Raspberry PIs, do we want to consider a solution that doesn't require their use?

**Comment by orrious** [comment]
Digital Outputs:
24 minimum
5v or voltage selectable
optically isolated?
PWM is it required or nice to have?

**Comment by orrious** [comment]
Analog Inputs: (DAC)
Thermistors, tank level sensors, water level sensors, etc will use these.
8 minimum
0-5V range
Jumper for 10k to VCC to enable Thermistor voltage divider.

**Comment by orrious** [comment]
Analog Output:
Please post your requirements

**Comment by orrious** [comment]
Digital Input:
I only use 1 right now, how many should we include?
Optically Isolated to allow any input voltage?

**Comment by orrious** [comment]
Pulse counters:
2 - minimum 
This is the hardest part of the entire ask as it requires a MCU or multiple ABLIC S-35770

**Comment by rstrouse** [comment]
Honestly, I think that you should be thinking linux microprocessor for the main controller and separate bus topologies for internal equipment.  I2c or SPI would be best for localized I/O simply because it is robust and easy to implement without being bound to the processor timing.  In an ideal world Raspberry Pi makes so much sense since there is so much available for things like web servers, ethernet, USB, data servers, ...etc.   

If you want to support a broad swath of equipment out there you will need at least two separate RS485 ports.  MODBUS in the traditional sense is not really viable here since none of the external equipment uses standard CRC or headers.  Heck some of them are using protocols that smell a lot like TTL with AT commands and 0x0D/0x0A on an RS485 bus.  So think RS485 and abandon the idea of MODBUS over RS485.

I assume that some of this is due to the fact that CRC lookups are not required with latch based protocols.  Especially when you are in control of the message length.  Much simpler checksums are in order as well, because the slave addressing is specific to the manufacturers and the localized equipment.  Heck some of the equipment is interrupting cow where it will pull the line high and spit out an indiscriminate message from the slaves as if it were upnp over RS485.  The expectation is that it can get collided but sooner or later the intended receiver will get the message after periodic resends.

As for other I/O I don't really see that much need for pulse counters or PWM.   In the case of digital inputs de-bounce on a hardware interrupt is easier to implement and more versatile overall.  With PWM, you will be hard pressed to find any pool equipment that needs it.  As an input strategy you are more likely to find 4-20mA signaling and even then this will be limited to pulse counters for things like water meters or adapted industrial equipment (think SCADA).   PWM directed output in my opinion is too localized for any real benefit.  We aren't driving triacs or LEDs.  If we were I would say stop it and think DMX or WS2812 to fix the pool lighting disaster once and for all.

In the end, I see four primary I/O subsystems.  First, there is analog inputs for which at a minimum you should have one 4 channel ADC like an ADS1115.  This could even be a single analog input that is multiplexed.  Heck I just built a board for my soil moisture sensing that uses a MUX for 15 analog inputs on ESP8266 to give 15 round robins + an internal VCC reading.

Next you will need some digital inputs.  These come into play for things like spa side remotes, flow switches, and even hardwired functions on the pad.   I would think you could get by with 2 of these but for spa side remotes you need at least 4 since this seems to be the magic number that the manufacturers settled on.   Bear in mind the more sophisticated remotes use the interrupting cow method over RS485.  So far I have witnessed only Jandy using probe messages for such equipment which accounts for why it seems so unresponsive.

As for digital outputs there are really two categories.  These include Circuit relays and Valve relays.   Circuit relays are typically on/off (NO/OFF) or single pole arrangements.  I suppose you could use a darlington array or CMOS switching circuits for this to eliminate the need for relays on the board itself.  In the end you do not want latching style for this.  If the watchdog is tripped or the power is lost it should result in open.  At a minimum there will be a need for at least 10 digital output relay drivers.

Valve relays must be powered with at least one leg at all times. If not, newer valve designs such as IntelliValve will not work.  Internally they get their power from a bridge rectifier and capacitor arrangement that gives consistent DC power to the innards when the legs are switched.  For this I am not sure there is any way around mechanical relays.  I don't think SSRs or CMOS switching belong in the valve driver.  They add failure points that could destroy the valve should both legs get powered at the same time for more than a few ms.

Finally, if you use 24vac as the main power transformer and a switching PSU from that to get 5vdc everything else can be driven either through a bridge rectifier or linear regulator to power all the electronics and peripherals.

Just a bit of my two cents on the subject.

**Reply by orrious** [reply]
> I think that you should be thinking linux microprocessor

Agree.  Where is that physically; on the pad or remote?  On the pad is easy, but hw is more difficult to find.  

Divorce the CPU from the pad and it would allow me to run njs-pC with all of my other containers.  We then need a reliable transport between the pad and CPU.  WiFi is not reliable enough, IMO.

**Reply by rstrouse** [reply]
I would agree.  That is why njsPC and dashPanel are separate server processes.  This means that they are not bound to hardware and can be run in some other location than the pad.   So you can spin up njsPC and dashPanel wherever you like.  REM was originally developed to bang protocol messages at IntelliValve in the hopes of talking some sense into it.  It was used to simply flip relays and interrupt the RS485 comms as needed to get the valve to wake up.

If something can be wired with ethernet it will be wired with ethernet in my house.  This means that no wall, ceiling, or concrete pad is safe from demolition to get the job done.  Heck if I were Geppetto I probably would have put the strings back on Pinocchio.  While WiFi is unreliable, I do find that TCP is highly reliable and robust.  More so even than RS485 which can run a close second with proper ACK.

Radios are by their very nature unreliable and the more of them that exist in the area the less reliable they become.  The air really does not have waves rather it is a choppy mess of turbulence.   Microservices and sockets are a good blend of performance, reliability, and error checking so long as you don't add protocols on top of them that kill atomic transactions.  That is why MQTT confounds me.

<mqtt:rant>Lets take a half-open socket and create a broker.  Then in an ill-fated attempt to ensure delivery, create a subsystem that delivers the message to another half-open socket without any indication that delivery took place or not</mqtt:rant>.  Good as a chat room but not great for control systems should the ear or mouth be inop.

**Reply by orrious** [reply]
I agree with everything you've stated.  So lets work with the foundation that everything must be hard wired and we will not support WiFi or other RF as the transport.

I would like to discuss the pros and cons of running ethernet vs another protocol atop of that wire.  Ethernet is extremely reliable yet prone to electrostatic spikes.  The MCU power and libraries needed to leverage it and then the cost of the actual hardware.  

vs a dedicate RS-485 bus for REM.  I.e that same cat5e cable is 4 twisted pairs, which means it could support 4 different RS485 buses.  Then each of your docker containers on the CPU could mount whichever RS485 bus it wanted.  i.e bus-a: njs-pC <-> Pentair pumps / Chlorinator.  bus-b: rem <-> modbus to relay boards, gpio boards, etc.

**Reply by orrious** [reply]
For clarity sake, I'm not married to MODBUS atop of the dedicated RS485 bus.  I do see its merits and what continues to draw me to it are:
1.) Open source protocol and standardized.  ( I don't have to create some new protocol that will incur unneeded tech debt).
2.) Libraries readily available for all languages
3.) Off the shelf components will just work
4.) MODBUS works across TCP too!
5.) Off the shelf extenders, tranceivers, media converters, etc
6.) RS485 chips are cheap and requires little overhead.

**Reply by rstrouse** [reply]
The protocol is separate from the transport TCP, RS485, 4-20mA... etc all can be radioed.  WiFi however is where folks tend to end up and unfortunately, it is never implemented in an open field with line of sight to all devices.  The other part that makes this rough is that folks seem to think that 2.4Ghz light switches, toasters, doorbells, cameras, refrigerators, TVs, and everything else is a great idea.  It's like spending money because you still have checks in your checkbook.  Most consumer WiFi routers cannot handle that many connections.  Heck I have over 200 ZWave devices in my home that would bring most WiFi routers to their knees.  Yet the marketeers of said smart home devices keep turning to WiFi.   Hell the WiFi 6 variants of 802 address only the packet routing issue and not the channel hopping.  It will be better for these devices but still not rising to the reliability level of a wired device.

As far as the interference goes there is a much more robust ECC for TCP than all the other protocols we have discussed,  This is complete with guaranteed delivery and receipt notification.  Throughput is very high because every implementation is SoC with dedicated buffers, signal processing... etc.

If you are referring to protection from lightning strikes anything with a copper wire is susceptible.  So some level of surge protection is in order for any home run into the house.   Lightning does not seek out the Ethernet protocol like tornados seek out mobile homes.  It is simply looking for path to ground.  If you ran RS485 vs Ethernet either one could be part of that path and TVS diodes offer no relief from a lightning strike.  Real lightning protection for either media costs just about the same and should be considered for any comms lines outside the home.

Now after giving WiFi its proper beatdown there are conditions where it will work.  If you have great connectivity -40dBm or greater without dropouts I say go for it.  However, don't expect anything to play nice with 30 wifi devices or other radio sources in the 2.4Ghz range around.

As for RS485 driving things most of them sit atop a 16550 uart either in software or hardware are typically CPU bound.   For this implementation I don't think it is a problem but the reason we see it widely used in the pool equipment automation space is because it was a longer range alternative to RS232 at the time for serial comms.  And RS422 required more sophisticated modems for multi-drop comms due to the fact that it is full-duplex and it needs termination at short distances.  If RS422 had become the standard there would have been comms issues when adding more than one device to the bus.  While it is differential voltage communication, the reference is highly dependent on termination in the right place.

The point is that its reason for RS485 existence is that it is good enough for intelligent slaves without distance limits, but I would not consider it an overall control bus.  Even CAN and LIN in autos are being replaced by 802 protocols these days.  Not because they need multi-gig transfers but primarily because of the SoC availability and robustness of the protocols.  

I guess what I am trying to say is that microservices on pretty much any controller are probably the best bet these days.  They are easy to develop, debug, and transport no matter the media or the overall route.

**Reply by orrious** [reply]
It's easy for an auto manufacturer to dictate which protocols are going to be running over their closed network.  Going with an ethernet transport "in the wild" will have you perpetually integrating the next device's API.  That's the trade off for not requiring special hardware at the CPU, but as long as we make that decision eyes wide open, I'm good with it too.

[RJ45 Port Ethernet Control Module with 16-Channel Relay LAN WAN Network Web Server Network Ethernet Relay Control](https://www.amazon.com/Ethernet-Control-Module-16-Channel-Network/dp/B08J6XFS1B/ref=asc_df_B08J6XFS1B)

**Reply by arieltools** [reply]
Another:
[Ethernet RJ45 TCP/IP Relay Remote Control Controller Network Web Server 8 Channels Relay Module Board 250V/AC 10A with Case
](https://www.amazon.com/dp/B076CQ4CYL?ref_=cm_sw_r_ud_dp_MAA73EHZ4DZQBXP64372)

**Comment by orrious** [comment]
What about a REM gateway board?

MCU on board
Ethernet - PoE powered
2 channels of RS485
2 channels of I2C
1 SPI, 4 GPIOs for SPI selection

njs-pC could use one RS485 and REM the rest. Open a TCP socket to each device and stream it via TCP.  This would allow you to support all of your existing external devices as well and run your compute in docker containers on proper HW.

**Reply by rstrouse** [reply]
That is what I thought you were referring to.  If there was a version of the Expressif chip with an Ethernet port it would already exist.  Sadly, they are focused on WiFi.  A couple of days on the C++ compiler and it would have been a reality.

**Reply by orrious** [reply]
Just found this...

https://wesp32.com/

https://www.amazon.com/Silicognition-wESP32-Isolated-Ethernet-802-3at/dp/B084RV91NQ

https://cdn.hackaday.io/files/853893653282976/wESP32-1.pdf

**Reply by orrious** [reply]
From what I can tell there is 2x UARTS we could use for RS485, 1x SPI, and 2x I2C buses.  I think that's what I just asked for.. It will be here Wednesday.

**Reply by rstrouse** [reply]
Interesting. I couldn't wrap my head around it at first because I didn't see a CH340 on it but you will likely need the programmer module as well.  Once you get some software on it you can upload fw to it remotely.  I have written a module that does such a thing from a firmware file on a central server for my ESP8266 stuff.

**Reply by orrious** [reply]
I'm going to connect a MAX485 module I have to to UART1 and IO2.  MAX485 requires a UART and GPIO for enable.

I have no need for SPI and will sacrifice those pins for an additional I2C bus.

I think each device should be on it's own TCP port / socket.  I'm not going to worry about authentication at first.  I've already found a couple other projects that bridge UART / MAX485 to TCP.  Hopefully there is something for I2C as well.

**Reply by rstrouse** [reply]
I wouldn’t put each device on its own port.  Use 1 socket for each client and open a listener port for microsevice requests.  njsPC will send you a microservice latch and maintain a client agent on its end.  Analog emits for sensors over the connected sockets is also ridiculously straight forward.  The only thing missing on the platform is SEH so it will require some care in buffer management.

You will find that json payloads can be a godsend and allow some flexibility on the expected payload at the expense of body parsing and formatting.  But anybody who has sent binary RPC and tried to keep interface contracts will appreciate the robustness.

You can use that same listener for a web interface that can be used for firmware updates and configuration.  I used Littlefs on ESP8266.

**Reply by orrious** [reply]
Do we want more than one client to be able to connect to a hardware port? (i.e I2C or UART/RS485).  I was going to treat this almost as a traditional terminal server.  Once there is a client connected to a particular port, it's locked and no additional clients may connect.

**Reply by rstrouse** [reply]
I think that is a judgement call.  If this were on another platform with socket filters I might say it could be a room/namespace subscription.  A proxy for a serial port might actually be easier because there is sooooo much code behind decoding and encoding rs485 packets.  Buffer and stop bits are easy enough but identifying slaves, message actions, and unpacking bits is probably a bit past the pail on that platform.

I assume you are thinking c++ and not Lua.

**Reply by orrious** [reply]
Straight C, no ++

For rs485, it's already on a UART, with clearly defined start and stop bits.  Sending that serial data over TCP is trivial.  I'm not so worried about this.  @rstrouse  What would be helpful in testing this is a virtual pump.  Docker container that will attach to USB rs485 adapter and mimic an Intelliflow pump.  It will allow me to validate the gateway I'm creating.

I2C is a different beast all together.  While layers 1 and 2 are clearly defined. (Physical media and data link (how things are addressed on the bus, etc)) However what has not been defined and is now the wild west of I2C is how each board communicates.  Yes there is a board ID, but after that there are no "bumpers" to keep all the devices talking the same.

What I am trying to get at is:  There is not going to be an easy way to proxy all the I2C traffic back and forth.  I see two paths forward for just making an I2C / TCP gateway:
1.) We would literally need to send each Byte back and forth and let REM decide when the I2C "sentence" was done.  This is really prone to errors as each Byte will need to be pristine.  Any retransmissions could significantly slow things down.
2.) Program the ESP32 for each board you wanted to use so it knows the I2C protocol being used and can decode it locally.  Then you can send the whole "sentence" in one shot.  

The other alternative is to do all the I2C on the ESP locally and just put up websocket / mqtt to control it.  The ESP32 is more than powerful enough to run all of this and way more.  CPU runs at 240mhz.  That's 240 million instructions per second.  You could run a full TFT display via SPI, OTA, webserver, mqtt client, pulse counters and still have 3/4 of your cycles free.  These things are seriously a beast.

At the end of the day it comes down to support and where we want to put the development effort.  

I keep coming back to building my own hardware board.  The wESP is a good start but we are still going to need to interface things to it, like max485 and I2C boards, etc.  It's going to look like a breadboard was shoved in my panel with wires going all over the place. Not aesthetically pleasing, IMO.  So more than likely I'd end up creating some kind of daughter card to clean it up with on board MCP23017.  ESP32 is 3.3v so need level shifters and opto isolators, as well and I'm almost back to where I started with a Rpi.  

Which has me coming back full circle to, should we even use ESP32?  I don't want to be the only one supporting this solution.

**Reply by rstrouse** [reply]
I went round and round with possibilities for a hardware abstraction server.  That is the primary purpose of REM.  The reason why it is required is because when it comes to hardware there is no such thing as cross platform.  REM is simply a server process that allows for access to the hardware using a server to server topology.   I2c, SPI, 1-wire, and GPIO are not directly interfaced to njsPC.  In fact njsPC doesn't need to be the consumer.

We wanted njsPC to be as device agnostic as possible.   RS485 is the one area where hardware needed to be addressed so we made a subsystem for that so that it could communicate with a remote RS485 adapter over TCP.  If you install a networked RS485 adapter or stand up an RS485 proxy, njsPC will digest the RS485 conversation as if it were holding onto the wires itself.

We landed on Rpi simply because of all the OTS hardware available.  Unfortunately, when it comes to electronics there has been tremendous wilt in the availability of options.  Heck if one were to develop a board with binary switches in whatever form to switch circuits, three state relays for valves, and at least 4 ADC channels for sensors, you would have a very clean and kickass implementation.

For the binary drivers one could use transistor arrays, relays, or CMOS switches,  For the valves, just stick to mini electromechanical DPST relays.  If one were to pair that with some I/O extenders the number of supported valves and circuits are essentially endless.

Popping a microcontroller on the board with an I2c interface to microprocessor (like a pi) to do the abstraction, configure the device, and serve device data over TCP with JSON payloads would allow processes like HA integration, schedule control, OCP slave, and any consumer of the device state/config to be split to wherever you wanted them to run.  Conversely, it could also run where all of it is installed on the microprocessor with the server to server comms running through the loopback.

I hear you on the support side of things.  When I originally got involved in this project my goal was to add platform support for IntelliCenter and open it up so I could use my HA then fade away into the darkness.  Unfortunately, I became irritated by the fact that my pad full of IntelliValves didn't have RS485 support so REM was born.  After a SETI like project to see if there was any intelligent life on those valves, it was determined that we are alone in that universe.  But by that time I had seen the wisdom of @tagyoureit's vision and had to see it through.

Then came chemistry control, complete with a whole community of folks looking to live the pool life without doting over it.  So kicking and screaming @tagyoureit and I got to work making a top notch chem controller and the list of supported equipment grew.  Nixie was simply the next logical step.

You do know that you can install REM on that pi zero and you will be done.   REM will stand up to control the I2C, RS485, and SPI.  Then you simply run the njsPC and dashPanel processes wherever you like.  Hell you can run them in the cloud with proper routing.  The only reason REM needs to be running at the pad is so it can touch the hardware.  If you want to build a driver interface with sensor support, relays, and RS485 simply include them in REM and you are done.  That is kinda what the MEGA-BAS is.   If you install REM locally on a non-pi platform you can add a BAS then see all the inputs and outputs configuration available.
![image](https://user-images.githubusercontent.com/47839015/176234375-b40b6a57-d3b3-4a5a-9c6a-bfe376d5e888.png)

**Reply by orrious** [reply]
Do we have a relationship with anyone at Sequent?  What do you think the chances of them building a board(s) for us?  The megabas with something like a "16 Universal Inputs" but as 32 digital output at 5v or selectable via jumpers.

**Reply by orrious** [reply]
When using the megabas or other hats, how are you connecting your I2C wires for Atlas probes?  Looking for something with terminal screws.

**Reply by johnny2678** [reply]
> Then you simply run the njsPC and dashPanel processes wherever you like.

As of last week, my updated software/hardware stack (minus containers for MQTT, Influx, & Granfa) is 100% on pad.  But prior to this, I only ran REM on the pad while njspc/dash were run indoors.  Been running REM since it came out almost 2 years ago and Wifi connection was solid and that was through block wall.

**Reply by johnny2678** [reply]
> how are you connecting your I2C wires for Atlas probes?

The tentacle product line from whitebox - https://www.whiteboxes.ch/product-category/boards/whitebox-isolators/

I'm using the T3 on my pi4

**Comment by orrious** [comment]
Rpi zero with [PoE Ethernet/USB HUB HAT](https://www.amazon.com/gp/product/B092W1Z8RW) as pad CPU 

I've been giving this additional thought.  While Rpi 3B+/4 aren't that readily available right now, hopefully that resolves itself.  Rpi Zeros are a little easier to find.  Rpi is also readily supported by our community and I won't be the only one doing it.  (Notice a theme here?) 

I see your motivation to leverage Ethernet as the transport, no additional HW needed at the CPU.  This overcomes the issues I currently have with my setup and Rpi. (I'm currently wireless.  I planned to put everything at the pad to resolve instability, but hard wiring does the same thing).

Strip everything out of REM and njs-pC that doesn't need to be on the pad.  I.e keep all the RS485 / I2C / SPI there and then do everything else on Linux CPU.  Then if you have a RPI4 and want to run it all on a single Rpi, you can do that too or break it up on low end Rpi.  Then you don't loose any work / compatibility with all the other boards you've already integrated with.

**Comment by rstrouse** [comment]
> Do we have a relationship with anyone at Sequent? What do you think the chances of them building a board(s) for us? The megabas with something like a "16 Universal Inputs" but as 32 digital output at 5v or selectable via jumpers.

Yes we do however, I think Sequent struggling to keep their current product lines in stock.  Everything is in short supply and when you can't get a microcontroller for the current stock, I couldn't imagine starting another product line.   They go out of stock frequently on most of the current line.   Something to note you can get up to 64 relays by stacking the 8 relay card.  After that you can run multiple REMs for another 64.

If you just want to expand the I/O for your own relay control you can simply get some MCP23017 breakouts and address them to I2c.  Then select the multi-relay device type in REM and choose the MCP23017 based controller.  This will switch the GPIO outputs on that board.  All of the ones I have seen allow stacking of that those as well for a range of 7 addresses on the A0-3 pins.  That is up to 112 GPIO.

**Reply by orrious** [reply]
I settled on just that this morning. Now waiting for everything to get here.  The only thing I'm not excited about is the level shifter I'll need to go between Rpi and MCP23017 to run at 5v.

Rpi 3B+
PoE hat (yes I really enjoy PoE)
megabas
[DIN Rail Mount MCP23017 32-Channel GPIO Expander with I2C](https://www.aliexpress.com/item/3256803335438797.html?gatewayAdapt=4itemAdapt)
Whitebox T3 and move all my EZO's over to that
https://www.amazon.com/Electronics-Salon-Terminal-Breakout-Module-Raspberry/dp/B01M27459S might put the level shifter on this.

**Comment by orrious** [comment]
What are folks using for GPIO Expander.  The [DIN Rail Mount MCP23017 32-Channel GPIO Expander with I2C](https://www.aliexpress.com/item/3256803335438797.html?gatewayAdapt=4itemAdapt) hasn't even shipped yet

**Comment by rstrouse** [comment]
There are lots of MCP23017 breakouts that should do the trick. 
https://www.amazon.com/MCP23017-IO-Expansion-Board-Interface/dp/B082MMRNM4/ref=sr_1_17?gclid=EAIaIQobChMIpMiI-JHn-AIVmTytBh3hhQCuEAMYAiAAEgIvgvD_BwE&hvadid=570464375448&hvdev=c&hvlocphy=9032516&hvnetw=g&hvqmt=e&hvrand=9341356565886681629&hvtargid=kwd-301289350522&hydadcr=24333_13533857&keywords=gpio+expander+i2c&qid=1657209619&sr=8-17

You can also use PCF8574 but that was programmed to the data sheet and to my knowledge has never seen an actual chip at the other end of it.
https://www.amazon.com/Comimark-PCF8574-PCF8574T-Expander-Raspberry/dp/B07X3KWQZ7/ref=sr_1_3?gclid=EAIaIQobChMIpMiI-JHn-AIVmTytBh3hhQCuEAMYAiAAEgIvgvD_BwE&hvadid=570464375448&hvdev=c&hvlocphy=9032516&hvnetw=g&hvqmt=e&hvrand=9341356565886681629&hvtargid=kwd-301289350522&hydadcr=24333_13533857&keywords=gpio+expander+i2c&qid=1657210228&sr=8-3

**Reply by orrious** [reply]
I've only been able to find one MCP23017 board with screw terminals on it, have had it on order for over a week, and it's yet to ship.  This is the last piece that is preventing me from moving to REM.  In the mean time I'll go with the solution you've provided.  Just going to be grungy.  Can you share your contact at Sequent?  At this point I would over pay for a 32 channel dual MCP23017 hat.

**Reply by rstrouse** [reply]
You can send an e-mail to support@sequentmicrosystems.com and ask for Alex or Mihai.  You know a soldering iron is your enemy until it is not anymore.

**Reply by rstrouse** [reply]
I broke down and bought these.  If you want it to have screw terminals you can simply get some 2.5mm pitch screw terminals and solder them on.  I'll need to add the register support to REM for it but it is just as PCF8574 with two 8bit R/W registers.  Interestingly the breakout has a level shifter built in.

https://www.amazon.com/gp/product/B09DFWS722/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&th=1

**Reply by orrious** [reply]
Soldering is not the problem, I'm proficient at full SMD rework and assembly.  The problem is I'm super OCD and like a clean look.  

https://www.amazon.com/gp/product/B0B2P6X6HD

I had ordered these earlier for the: https://www.amazon.com/MCP23017-IO-Expansion-Board-Interface/dp/B082MMRNM4

but the one you just found has built in level shifter.  So it's now preferred.

**Reply by rstrouse** [reply]
Yeah that is why I decided to try it.  It is an NXP chip not an MCP so I will see how I like it and it doesn't seem to use a command byte.  It looks like it only has 2 read and 2 write registers.  Then again MicroChip produces sooooo much better data sheets for folks that are soft in the middle.

**Reply by orrious** [reply]
Reading the data sheet: https://www.ti.com/lit/ds/symlink/pcf8575.pdf

You can use any of the 16 GPIOs for either input or output.  Where is differs from the MCP23017 is that it does not break up the 16 GPIOs into two separate banks.  The MCP23017 has an interrupt for each bank that must be ran back to the RPI/MCU GPIO, whereas the PCF8575 has a single interrupt for all 16 GPIOs.  

Keep in mind this is only needed if you use the MCP23017 or PCF8575 as input devices.  The interrupts are how the expanded chips tell the MCU to come check the GPIOs because something has changed.

**Reply by rstrouse** [reply]
Yeah but that interrupt is only when all of the GPIOs are configured as inputs.  In reading the sheet I finally came to the conclusion that there is no configuration register and it only reads and writes raw I2c.  I am still unsure how one would configure the physical pin as an output or input without a config register though.  Maybe it becomes an output as soon as you write to it but it is always an input.  I guess I will see when it shows up.

This is the level of documentation for for read and write registers.  According to their chart the read address is 8 bits and the write address is 8 bits and they are sequential (maybe they are selector registers).  Mathematically you don't get 16 states from that.  Or the more likely case they are counting on the R/W bit on a 7-bit addressing scheme to determine the register pair.  Either way why build such a confusing register chart?
![image](https://user-images.githubusercontent.com/47839015/178046066-9e301d80-8900-443d-8e7e-fa3b539ec4f7.png)

**Reply by orrious** [reply]
A0 - A2 are how you set the I2C board ID, 3 bits will give you 2^3=8 possible combinations.  Each device on the I2C bus gets its own board ID from 0-255.  If I'm understanding this correctly, this chip is occupying two board IDs.  The odd board ID is for reading from the GPIO and the even for setting an output.  

If you leave A0-A2 low, then you would use board ID 65 to read an input and board ID 64 to write an output.  Interesting way of doing it.

**Reply by orrious** [reply]
Btw, the interrupt back to the MCU is used if any of that banks GPIOs are set to input.  You can put both banks of an MCP on the same MCUs GPIO, but then you will need to go check both banks for changes to the input.

**Reply by rstrouse** [reply]
Yes I know how to set the I2c address.  It is the internal register/command byte that appears to not be used in this instance.  Typically the first bit after receipt of the ACK starts the register/command reference.  In this instance it appears that either a command reference for the read is base address + 1 or it is simply looking at the 7bit i2c address and determining which register to reference off the write bit.  I assume the latter.  I suppose this works because there are no configuration registers for things like the pu/d, latches, polarity or input triggers.

![image](https://user-images.githubusercontent.com/47839015/178079527-843dc8b0-de4a-4370-91ea-4140f038804b.png)



I am a little bit confused as to how it knows whether each pin is an input or output.  Maybe that is this quasi-bidirectional stuff it is talking about.

**Reply by rstrouse** [reply]
Do you have any ideas regarding whether it latches states.  Very bizarre that I write bytes to the P0x then when I read it back it appears to float.  According to the docs everything is supposed to be pulled up so I would have expected a different read.

**Reply by rstrouse** [reply]
Well after scratching my head on how this should work I figured it all out.  You must perform your own pullups or pulldowns externally from the chip.  The pullup is only inrush to trigger the pin.  It will fade back to floating.

As for the protocol it is raw I2c and uses the R/W bit to determine which command register it is looking at.   So it really is not SMBus in the sense that it uses a command byte to locate the data.  The data sheet talks about an 8 bit address on a protocol that only supports 7, 9, or 11.  The R/W bit is reserved but they documented it as an 8 bit address to get read and write access to both banks.

So as soon as I pulled a couple of the pins down it all made sense.  I suppose if I had not implemented the MCP series I might not have assumed that the latch was automatic using internal p u/d resistors.   Either way you can pull REM and the PCF8575 and PCF8574 (single bank) have been checked against hardware.  Please note that you will want to pull the pins down if you are using them for output or invert them and pull them up.

**Reply by orrious** [reply]
I2c address is always 8 bit.  (Decimal: 0-255, hex 0x0-0xFF, or binary 00000000-11111111). The read/write bit is the 1s, least significant bit in the board address, which is 2^1 and signifies if the number is even or odd.  The 4 most significant bits are hard coded to Dec 64, Hex 0x40, binary 0100xxxx.  This is why the board address start at 64.    

From what I can surmise from the data sheet.  All outputs are overwritten ever time you send a write command.  When writing, you must send 3 byes each time.  1 byte board address (LSB 0), 2nd byte P07-P00 (first bank), 3rd byte P17-P10 (second bank). You have to set the output to high for a channel before you can use it as an input.  **If a high is applied externally to an I/O that has been written earlier to low, a large current will flow to GND**.   This means you have to be very careful and know what you are attaching to this device as you can potentially sink too much current across this device and smoke it.

I don't know if you can treat this like a normal MCU GPIO where you write and then go back to read to validate that it worked properly.
"Before reading from the PCF8575, all ports desired as input should be set to logic 1. To read from the ports
(input mode), the master first addresses the slave device, setting the last bit of the byte containing the slave
address to logic 1. The data bytes that follow on the SDA are the values on the ports. If the data on the input port
changes faster than the master can read, this data may be lost.



I didn't catch what you meant about latching state?  The outputs should automatically latch to the last state you set.  Is that working?

If you are talking about INT:   INT appears to latch until the PCF is read.  INT is to be tied to your MCU INT for the purposes of triggering an interrupt on the MCU.  When the MCU receives the interrupt it should immediately read the PCF.   

"By sending an interrupt signal on this line, the remote I/O can inform the microcontroller if there is incoming data on its ports, without having to communicate via the I2C us. Thus, the PCF8575 can remain a simple slave device."

**Reply by rstrouse** [reply]
Strictly speaking I2c addresses are always 7 or 10 bits by spec.  However, the bit immediately following the address is the R/W bit.  For instance if you address the board to 0x20 and poll the slaves on the bus it will not respond to 0x21.  In fact you can address another slave on 0x21 and still read and write to the PCF8575 on 0x20.  This is a distinction and it does have a difference.

Since this device only has two registers, one for reading and one for writing, they eliminated the need for a command byte on the protocol.  However, when reading I am not communicating with a slave at 0x21 not using an SMBus command byte no matter how one slices it.  There are many devices that paint outside the lines like this but the docs spun me out for a bit.

> I didn't catch what you meant about latching state? The outputs should automatically latch to the last state you set. Is that working?

Yes this is working but the pins actually float without a default state.  So on boot it sets the pin high and fades then if you set the pin high it will set the pin high for about 750ms then fade off after the inrush current when nothing is connected to the pin.  If floats just like a regular GPIO in that sense.  What confused me was that I assumed that as soon as I wrote the pin high it was going into output mode.  However, reading it back as the pin floated gave random 0-1 responses on the bit.  If you are actually driving something on the pin it will not exhibit this behavior.

My guess is that is what they mean by quasi-bidirectional in that it is always an input pin and you do not set the pin up before an operation.  So if you do something stupid like drive the pin high with a hw configuration that would traditionally be for reading you can do some damage with inrush current.  Simply adding a current limiting resistor on the pin alleviates this potential disaster though.

Where this differs from the traditional GPIO is that there are internal protection and p u/d resistors at play.  So driving an LED with the pin works like a champ because the pin is current limited and its state is either pulled up or down depending on the configuration of the circuit.  So yes this works as expected once the pin is driving something as output or actually reading something as input.

I was interpreting the read from the chip with floating values as me not reading the proper registers after write.  What was confounding is that I put some code in to read the register back immediately after write and it was working but a subsequent read 1 second later resulted in random states for the pin.  The pin was simply floating but because I wasn't sure yet of what they meant by a read register on I2c slave address 0x21.  (I have the chip addressed to 0x20).  The first clue should have been, when A0-2 are low it is not on I2c address 0x40, I2c detects it on 0x20 which is exactly where it is at.

The RP gpio and MCP use latch registers to keep all of this state in order so if something does get connected to the output pin will be at the state you think it should be.  I wouldn't consider this an oversight it is just the behavior of the chip and if you aren't expecting something like that while coding 16 outputs that have nothing connected you might be confused by the results.  

Although I didn't include anything for inputs on the chip I did hook up a button and read the state witnessing the interrupt pin going high when the button was pressed.  Also reading the state worked as expected.  You have to be careful when writing output pins to make sure you never set any input pins high.
